core/opendaq/synchronization/include/opendaq/sync_component_impl.h

```diff
@@ -0,0 +1,308 @@
+{
+    OPENDAQ_PARAM_NOT_NULL(interfaces);
+    ListPtr<IPropertyObject> interfacesList = List<IPropertyObject>();
+
+    BaseObjectPtr interfacesValue;
+    StringPtr str = "Interfaces";
+    ErrCode err = this->getPropertyValue(str, &interfacesValue);
+    if (OPENDAQ_FAILED(err))
+        return err;
+
+    const auto InterfacesPtr = interfacesValue.asPtr<IPropertyObject>(true);
+    for (const auto& prop : InterfacesPtr.getAllProperties())
+    {
+        if (prop.getValueType() == ctObject)
+        {
+            BaseObjectPtr interfaceProperty;
+            err = InterfacesPtr->getPropertyValue(prop.getName(), &interfaceProperty);
+            if (OPENDAQ_FAILED(err))
+                return err;
+
+            interfacesList.pushBack(interfaceProperty.detach());
+        }
+    }
+
+    *interfaces = interfacesList.detach();
+    return OPENDAQ_SUCCESS;
+}
+
+template <typename MainInterface, typename ... Interfaces>
+ErrCode GenericSyncComponentImpl<MainInterface, Interfaces...>::getInterfaceNames(IList** syncInterfaceNames)
+{
+    OPENDAQ_PARAM_NOT_NULL(syncInterfaceNames);
+    return daqTry([&]() {
+        *syncInterfaceNames = getTypedProperty<IList>("InterfaceNames").detach();
+        return OPENDAQ_SUCCESS;
+    });
+}
+
+template <typename MainInterface, typename ... Interfaces>
+ErrCode GenericSyncComponentImpl<MainInterface, Interfaces...>::addInterface(IPropertyObject* syncInterface)
+{
+    OPENDAQ_PARAM_NOT_NULL(syncInterface);
+
+    PropertyObjectPtr interfacePtr = syncInterface;
+
+    //TBD: Check if interface inherits from SyncInterfaceBase
+    StringPtr className = interfacePtr.getClassName();
+    if (className == "SyncInterfaceBase")
+    {
+        return this->makeErrorInfo(OPENDAQ_ERR_INVALID_ARGUMENT, "Allowed adding property objects that inherit from 'SyncInterfaceBase', but not 'SyncInterfaceBase' itself.");
+    }
+
+    auto typeManager = this->context.getTypeManager();
+    if (typeManager == nullptr)
+    {
+        return this->makeErrorInfo(OPENDAQ_ERR_ARGUMENT_NULL, "TypeManager is not assigned.");
+    }
+
+    TypePtr type;
+    ErrCode errCode = typeManager->getType(className, &type);
+    if (OPENDAQ_FAILED(errCode) || type == nullptr)
+    {
+        return this->makeErrorInfo(OPENDAQ_ERR_INVALID_ARGUMENT, fmt::format("Interface '{}' not found.", className));
+    }
+
+    if (auto objectClass = type.asPtrOrNull<IPropertyObjectClass>(true); objectClass.assigned())
+    {
+        auto parentName = objectClass.getParentName();
+        if (!parentName.assigned() || parentName != "SyncInterfaceBase")
+        {
+            return this->makeErrorInfo(OPENDAQ_ERR_INVALID_ARGUMENT, fmt::format("Interface '{}' does not inherit from 'SyncInterfaceBase'.", className));
+        }
+    }
+    else
+    {
+        return this->makeErrorInfo(OPENDAQ_ERR_INVALID_ARGUMENT, fmt::format("Interface '{}' is not IPropertyObjectClass", className));
+    }
```

Comment:
This should check recursively up the class inheritance chain for whether the base class is "SyncInterfaceBase"